core/predicate_check.go

```diff
@@ -0,0 +1,40 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/params"
+	"github.com/ava-labs/subnet-evm/precompile/contract"
+	"github.com/ava-labs/subnet-evm/precompile/modules"
+	"github.com/ava-labs/subnet-evm/utils"
+)
+
+// CheckPredicates checks that all precompile predicates are satisfied within the current [predicateContext] for [tx]
+func CheckPredicates(rules params.Rules, predicateContext *contract.PredicateContext, tx *types.Transaction) error {
+	precompileConfigs := rules.ActivePrecompiles
+	for _, accessTuple := range tx.AccessList() {
+		_, isPrecompile := precompileConfigs[accessTuple.Address]
+		if !isPrecompile {
+			continue
+		}
+
+		module, ok := modules.GetPrecompileModuleByAddress(accessTuple.Address)
+		if !ok {
+			return fmt.Errorf("predicate accessed precompile config under address %s with no registered module", accessTuple.Address)
+		}
+		predicater, ok := module.Contract.(contract.Predicater)
```

Comment:
How come we don't require `Predicater` implemented as part of `StatefulPrecompiledContract`? Is this so that some precompiles don't need to have a predicate function? I tend to like the predicate included in `StatefulPrecompiledContract` and not have the following `if !ok {continue}` check

Reply:
I prefer this so that it does not need to be implemented/filled out by every precompile even when it's only applicable to a few. Removing it from the standard template also makes it less likely that a custom precompile implementor will shoot themselves in the foot with this imo.

Open to changing it back if people feel strongly, but I prefer the optional interfaces here.

Reply:
I tend to prefer previous change but don't have strong preference. Think it helps with consistency amongst precompiles, and also think without prior context, it'd be hard to tell reading through the code that `StatefulPrecompiledContract` can implement the `Predicater`.

Reply:
I do agree that making the relationship between contract and predicater more hidden could prevent some incorrect predicate uses, so yea no strong preference.